File: HybridMeshRN/src/models/Device.ts
export interface Device {
  id: string;
  name: string;
  macAddress: string;
  ipAddress?: string;
  bluetoothAddress?: string;
  lastSeen: number;
  signalStrength: number; // RSSI for Bluetooth, signal level for Wi-Fi
  transportTypes: TransportType[];
  isOnline: boolean;
  publicKey?: string; // For encryption
  capabilities: DeviceCapability[];
}

export enum DeviceCapability {
  BLUETOOTH_LE = 'BLUETOOTH_LE',
  WIFI_DIRECT = 'WIFI_DIRECT',
  HOTSPOT = 'HOTSPOT',
  INTERNET_RELAY = 'INTERNET_RELAY',
  MESSAGE_STORE = 'MESSAGE_STORE'
}

export enum TransportType {
  BLUETOOTH = 'BLUETOOTH',
  WIFI_DIRECT = 'WIFI_DIRECT',
  HOTSPOT = 'HOTSPOT',
  INTERNET = 'INTERNET'
}

export const createDevice = (data: Partial<Device>): Device => ({
  id: data.id || generateUUID(),
  name: data.name || '',
  macAddress: data.macAddress || '',
  ipAddress: data.ipAddress,
  bluetoothAddress: data.bluetoothAddress,
  lastSeen: data.lastSeen || Date.now(),
  signalStrength: data.signalStrength || 0,
  transportTypes: data.transportTypes || [],
  isOnline: data.isOnline ?? true,
  publicKey: data.publicKey,
  capabilities: data.capabilities || []
});

const generateUUID = (): string => {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
    const r = Math.random() * 16 | 0;
    const v = c === 'x' ? r : (r & 0x3 | 0x8);
    return v.toString(16);
  });
};

File: app/src/main/java/com/hybridmesh/chat/data/model/Device.kt
package com.hybridmesh.chat.data.model

import androidx.room.Entity
import androidx.room.PrimaryKey
import java.util.UUID

@Entity(tableName = "devices")
data class Device(
    @PrimaryKey
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val macAddress: String,
    val ipAddress: String? = null,
    val bluetoothAddress: String? = null,
    val lastSeen: Long = System.currentTimeMillis(),
    val signalStrength: Int = 0, // RSSI for Bluetooth, signal level for Wi-Fi
    val transportTypes: List<TransportType> = emptyList(),
    val isOnline: Boolean = true,
    val publicKey: String? = null, // For encryption
    val capabilities: List<DeviceCapability> = emptyList()
)

enum class DeviceCapability {
    BLUETOOTH_LE,
    WIFI_DIRECT,
    HOTSPOT,
    INTERNET_RELAY,
    MESSAGE_STORE
}
